### PR TITLE
Microsoft.Data.Sqlite.Core issue with multiple Blob colums

### LINKED ISSUE #32747: Microsoft.Data.Sqlite issue with multiple Blob colums
## Microsoft.Data.Sqlite.Core issue with multiple Blob colums

Microsoft.Data.Sqlite.Core is not able to manage joined queries where
two blob fields belong from two different tables.

in the attached example:

```C#
command.CommandText = @"SELECT 
                            A.ID as AID,
                            A.DESCRIPTION as ADESC,
                            A.VALUE as AVALUE,
                            B.ID as BID,
                            B.DESCRIPTION as BDESC,
                            B.VALUE as BVALUE
                        FROM 
                            A JOIN B
                            ON B.FATHER_ID=A.ID ";
var reader = command.ExecuteReader();
while(reader.Read()) {
    Console.WriteLine($"A.ID={reader.GetInt32(0)} A.DESC={reader.GetString(1)} B.ID={reader.GetInt32(2)} B.DESC={reader.GetString(3)}");
    long abuffLength = reader.GetBytes(2, 0, null, 0, 0); //this line will force to cache rowid=1 from table A
    long bbuffLength = reader.GetBytes(5,0,null,0,0);     //this line will fail because rowid=1 for table B is invalid
}
```

The first line caches a rowid that is not usable for the second line.

Attached you can find a demo project that is able to reproduce the issue.

### Stack trace

```
Unhandled exception. Microsoft.Data.Sqlite.SqliteException (0x80004005): SQLite Error 1: 'no such rowid: 1'.
   at Microsoft.Data.Sqlite.SqliteException.ThrowExceptionForRC(Int32 rc, sqlite3 db)
   at Microsoft.Data.Sqlite.SqliteBlob..ctor(SqliteConnection connection, String databaseName, String tableName, String columnName, Int64 rowid, Boolean readOnly)
   at Microsoft.Data.Sqlite.SqliteDataRecord.GetStream(Int32 ordinal)
   at Microsoft.Data.Sqlite.SqliteDataRecord.GetBytes(Int32 ordinal, Int64 dataOffset, Byte[] buffer, Int32 bufferOffset, Int32 length)
   at Microsoft.Data.Sqlite.SqliteDataReader.GetBytes(Int32 ordinal, Int64 dataOffset, Byte[] buffer, Int32 bufferOffset, Int32 length)
   at Test.Program.Main(String[] args) in C:\Users\Franco\source\repos\SqliteBlobIssue\Program.cs:line 65
```

### Version information

Microsoft.Data.Sqlite version: HEAD, 8.0.0
Microsoft.Data.Sqlite.Core version: HEAD, 8.0.0
Target framework: NET 8.0
Operating system: Windows, Android, iOS

[SqliteBlobIssue.zip](https://github.com/dotnet/efcore/files/13863206/SqliteBlobIssue.zip)

### PR DESCRIPTION
**Pull request by francopettinari**
- [x ] I've read the guidelines for [contributing](CONTRIBUTING.md) and seen the [walkthrough](https://youtu.be/9OMxy1wal1s?t=1869)
- [ x] I've posted a comment on an issue with a detailed description of how I am planning to contribute and got approval from a member of the team
- [x ] The code builds and tests pass locally (also verified by our automated build checks)
- [ ] Commit messages follow this format:
```
        Summary of the changes
        - SqliteDataRecord was able to handle only one rowid
        - added RowIdInfo containing both Ordinal and Table to wich the ordinal is referred to
		- in GetStream function rowid is cached but with reference to the referenced table
		- now it is possible to handle sql satements with joins and multiple rowids and tables
		- Test case added and verified

Fixes #32747
```
- [x ] Tests for the changes have been added (for bug fixes / features)
- [x ] Code follows the same patterns and style as existing code in this repo

### CODE REVIEW

**Comment by francopettinari** [comment]
@dotnet-policy-service agree [company="{your company}"]

**Comment by francopettinari** [comment]
@dotnet-policy-service agree

**Comment by francopettinari** [comment]
> @francopettinari please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @dotnet-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @dotnet-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @dotnet-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement
> ### Contribution License Agreement
> This Contribution License Agreement ( **“Agreement”** ) is agreed to by the party signing below ( **“You”** ), and conveys certain license rights to the .NET Foundation ( **“.NET Foundation”** ) for Your contributions to .NET Foundation open source projects. This Agreement is effective as of the latest signature date below.
> 
> **1. Definitions.**
> 
> **“Code”** means the computer software code, whether in human-readable or machine-executable form, that is delivered by You to .NET Foundation under this Agreement.
> 
> **“Project”** means any of the projects owned or managed by .NET Foundation and offered under a license approved by the Open Source Initiative ([www.opensource.org](http://www.opensource.org)).
> 
> **“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any Project, including but not limited to communication on electronic mailing lists, source code control systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of discussing and improving that Project, but excluding communication that is conspicuously marked or otherwise designated in writing by You as “Not a Submission.”
> 
> **“Submission”** means the Code and any other copyrightable material Submitted by You, including any associated comments and documentation.
> 
> **2. Your Submission.** You must agree to the terms of this Agreement before making a Submission to any Project. This Agreement covers any and all Submissions that You, now or in the future (except as described in Section 4 below), Submit to any Project.
> 
> **3. Originality of Work.** You represent that each of Your Submissions is entirely Your original work. Should You wish to Submit materials that are not Your original work, You may Submit them separately to the Project if You (a) retain all copyright and license information that was in the materials as you received them, (b) in the description accompanying your Submission, include the phrase "Submission containing materials of a third party:" followed by the names of the third party and any licenses or other restrictions of which You are aware, and (c) follow any other instructions in the Project's written guidelines concerning Submissions.
> 
> **4. Your Employer.** References to “employer” in this Agreement include Your employer or anyone else for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your Submission is made in the course of Your work for an employer or Your employer has intellectual property rights in Your Submission by contract or applicable law, You must secure permission from Your employer to make the Submission before signing this Agreement. In that case, the term “You” in this Agreement will refer to You and the employer collectively. If You change employers in the future and desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement and secure permission from the new employer before Submitting those Submissions.
> 
> **5. Licenses.**
> 
> **a. Copyright License.** You grant .NET Foundation, and those who receive the Submission directly or indirectly from .NET Foundation, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third parties.
> 
> **b. Patent License.** You grant .NET Foundation, and those who receive the Submission directly or indirectly from .NET Foundation, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license under Your patent claims that are necessarily infringed by the Submission or the combination of the Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and import or otherwise dispose of the Submission alone or with the Project.
> 
> **c. Other Rights Reserved.** Each party reserves all rights not expressly granted in this Agreement. No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are granted by implication, exhaustion, estoppel or otherwise.
> 
> **6. Representations and Warranties.** You represent that You are legally entitled to grant the above licenses. You represent that each of Your Submissions is entirely Your original work (except as You may have disclosed under Section 3 ). You represent that You have secured permission from Your employer to make the Submission in cases where Your Submission is made in the course of Your work for Your employer or Your employer has intellectual property rights in Your Submission by contract or applicable law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You have the necessary authority to bind the listed employer to the obligations contained in this Agreement. You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES EXPRESSLY STATED IN SECTIONS 3, 4, AND 6 , THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
> 
> **7. Notice to .NET Foundation.** You agree to notify .NET Foundation in writing of any facts or circumstances of which You later become aware that would make Your representations in this Agreement inaccurate in any respect.
> 
> **8. Information about Submissions.** You agree that contributions to Projects and information about contributions may be maintained indefinitely and disclosed publicly, including Your name and other information that You submit with Your Submission.
> 
> **9. Governing Law/Jurisdiction.** This Agreement is governed by the laws of the State of Washington, and the parties consent to exclusive jurisdiction and venue in the federal courts sitting in King County, Washington, unless no federal subject matter jurisdiction exists, in which case the parties consent to exclusive jurisdiction and venue in the Superior Court of King County, Washington. The parties waive all defenses of lack of personal jurisdiction and forum non-conveniens.
> 
> **10. Entire Agreement/Assignment.** This Agreement is the entire agreement between the parties, and supersedes any and all prior agreements, understandings or communications, written or oral, between the parties relating to the subject matter hereof. This Agreement may be assigned by .NET Foundation.
> 
> _.NET Foundation dedicates this Contribution License Agreement to the public domain according to the Creative Commons CC0 1._

@dotnet-policy-service agree

**Comment by ErikEJ** [comment]
@francopettinari Where is the test?

**Comment by francopettinari** [comment]
@ErikEJ

SqliteDataReaderTest.GetBytes_works_streaming_join()

File: src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs [hunk 1]
```diff
@@ -14,16 +14,30 @@
 
 namespace Microsoft.Data.Sqlite
 {
+    
     internal class SqliteDataRecord : SqliteValueReader, IDisposable
     {
+        internal class RowIdInfo
```
Comment (ajcvickers) [inline src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs:20]: Any reason not to use a value tuple for this?

**Comment by ajcvickers** [comment]
@francopettinari Please add a test that fails before this change but passes after it.

File: src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs (hunk 1 above)
Comment (francopettinari) [reply src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs:20]: It was only for readability

**Comment by francopettinari** [comment]
> @francopettinari Please add a test that fails before this change but passes after it.

I already added a test.
class SqliteDataReaderTest, test function: GetBytes_works_streaming_join()

**Comment by ajcvickers** [comment]
@francopettinari The PR diff only shows changes to one file.

**Comment by francopettinari** [comment]
> @francopettinari The PR diff only shows changes to one file.

sorry, now pushed.

**Comment by francopettinari** [comment]
@ajcvickers I committed the test case. Should I do something else or comment/reply in onother way?

**Comment by ajcvickers** [comment]
@francopettinari No, I just need to get to reviewing it, but, of course, I have other things that need doing as well.

File: src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs [hunk 2]
```diff
@@ -368,22 +383,23 @@ public virtual Stream GetStream(int ordinal)
...
-                            _rowidOrdinal = i;
+                            rowIdForOrdinal = new RowIdInfo(i, tableName);
+                            RowIds.Add(rowidkey, rowIdForOrdinal);
                             break;
                         }
                     }
                 }
 
-                Debug.Assert(_rowidOrdinal.HasValue);
+                //Debug.Assert(rowIdForOrdinal!=null);
```
[the remark this answers is not in the document]
Comment (francopettinari) [reply src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs:393]: readding Debug.Assert(_rowidOrdinal.HasValue); will absolutely force to work with rowId.

[the remark this answers is not in the document]
Comment (francopettinari) [reply src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs:393]: In my example I have a join and for the second table it was not possible to define a rowid

[the remark this answers is not in the document]
Comment (francopettinari) [reply src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs]: Puppy error, sorry. Commented for initial checks and later committed. Not intentional. Sorry again!
I readded, relaunched the tests and finally committed.

**Review by ajcvickers — APPROVED** [review]

**Comment by ajcvickers** [comment]
Thanks for the contribution!

**Pull request merged**